src/ParallelAlgorithms/LinearSolver/Multigrid/ElementsAllocator.hpp

```diff
@@ -0,0 +1,181 @@
+#include "ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.hpp"
+#include "ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/ProtocolHelpers.hpp"
+#include "Utilities/System/ParallelInfo.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace LinearSolver::multigrid {
+
+/*!
+ * \brief A `Parallel::protocols::ArrayElementsAllocator` that creates array
+ * elements to cover the initial computational domain multiple times at
+ * different refinement levels, suitable for the
+ * `LinearSolver::multigrid::Multigrid` algorithm
+ *
+ * The `initial_element_ids` function and the option-created
+ * `domain::Tags::Domain` and `domain::Tags::InitialRefinementLevels` determine
+ * the initial set of element IDs in the domain. This is taken as the finest
+ * grid in the multigrid hierarchy. Coarser grids are determined by successively
+ * applying `LinearSolver::multigrid::coarsen`, up to
+ * `LinearSolver::multigrid::Tags::MaxLevels` grids.
+ *
+ * Array elements are created for all element IDs on all grids, meaning they all
+ * share the same parallel component, action list etc. Elements are connected to
+ * their neighbors _on the same grid_ by the
+ * `domain::Initialization::create_initial_element` function (this is
+ * independent of the multigrid code, the function is typically called in an
+ * initialization action). Elements are connected to their parent and children
+ * _across grids_ by the multigrid-tags set here and in the multigrid
+ * initialization actions (see `LinearSolver::multigrid::parent_id` and
+ * `LinearSolver::multigrid::child_ids`).
+ *
+ * This allocator also creates two sets of sections (see `Parallel::Section`):
+ * - `Parallel::Tags::Section<ElementArray,
+ *   LinearSolver::multigrid::Tags::MultigridLevel>`: One section per grid. All
+ *   elements are part of a section with this tag.
+ * - `Parallel::Tags::Section<ElementArray,
+ *   LinearSolver::multigrid::Tags::IsFinestGrid>`: A single section that holds
+ *   only the elements on the finest grid. Holds `std::nullopt` on all other
+ *   elements.
+ *
+ * The elements are distributed on processors using the
+ * `domain::BlockZCurveProcDistribution` for every grid independently.
+ */
+template <size_t Dim, typename OptionsGroup>
+struct ElementsAllocator
+    : tt::ConformsTo<Parallel::protocols::ArrayElementsAllocator> {
+  template <typename ElementArray>
+  using array_allocation_tags =
+      tmpl::list<domain::Tags::InitialRefinementLevels<Dim>,
+                 Tags::ChildrenRefinementLevels<Dim>,
+                 Tags::ParentRefinementLevels<Dim>,
+                 Parallel::Tags::Section<ElementArray, Tags::MultigridLevel>,
+                 Parallel::Tags::Section<ElementArray, Tags::IsFinestGrid>>;
+
+  template <typename ElementArray, typename Metavariables,
+            typename... InitializationTags>
+  static void apply(Parallel::CProxy_GlobalCache<Metavariables>& global_cache,
```

Comment:
yea it would be nice to unit-test the array allocations. I opened https://github.com/sxs-collaboration/spectre/issues/3373 to keep track.

This particular code is tested in the `Test_MultigridAlgorithm` executable test.